app/src/org/commcare/android/util/SigningUtil.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    // given a text message, parse out the cruft and return the raw Base64 bytes
+    public static byte[] getBytesFromSMS(String newMessage) throws Exception{
+        return Base64.decode(parseMessage(newMessage));
+    }
+    // given a text message, parse out the cruft and return the raw String
+    private static String parseAndDecodeSMS(String newMessage) throws Exception {
+        String parsed = parseMessage(newMessage);
+        return decodeEncodedSMS(parsed);
+    }
+
+    // given a text message, parse out the cruft and return
+    private static String parseMessage(String newMessage){
+        String parsed = newMessage.substring(newMessage.indexOf(GlobalConstants.SMS_INSTALL_KEY_STRING) +
+                GlobalConstants.SMS_INSTALL_KEY_STRING.length() + 1);
+        return parsed;
+    }
+
+    private static boolean verifySMS(String message, byte[] signature){
+        String keyString = GlobalConstants.CCHQ_PUBLIC_KEY;
+        return SigningUtil.verifyMessageSignatureHelper(keyString, message, signature);
+    }
+    /**
+     *
+     * @param publicKeyString the known public key of CCHQ
+     * @param message the message content
+     * @param messageSignature the signature generated by HQ with its private key and the message content
+     * @return whether or not the message was verified to be sent with HQ's private key
+     */
+    public static boolean verifyMessageSignatureHelper(String publicKeyString, String message, byte[] messageSignature) {
+        try {
+            PublicKey publicKey = getPublicKey(publicKeyString);
+            return verifyMessageSignature(publicKey, message, messageSignature);
+        } catch (Exception e) {
+            // a bunch of exceptions can be thrown from the crypto methods. I mostly think we just
+            // care that we couldn't verify it
+            e.printStackTrace();
+        }
+        return false;
+    }
+    // convert from a key string to a PublicKey object
+    private static PublicKey getPublicKey(String key) throws Exception{
+        byte[] derPublicKey = Base64.decode(key);
+        X509EncodedKeySpec spec = new X509EncodedKeySpec(derPublicKey);
+        KeyFactory kf = KeyFactory.getInstance("RSA");
+        return kf.generatePublic(spec);
+    }
+
+    private static boolean verifyMessageSignature(PublicKey publicKey, String messageString, byte[] signature) throws SignatureException, NoSuchAlgorithmException, Base64DecoderException, InvalidKeyException {
+        Signature sign = Signature.getInstance("SHA256withRSA/PSS", new BouncyCastleProvider());
+        byte[] message = messageString.getBytes();
+        sign.initVerify(publicKey);
+        sign.update(message);
+        return sign.verify(signature);
+    }
+
+    private static String decodeEncodedSMS(String text) throws  SignatureException{
+        String decodedMessage = null;
+        try {
+            decodedMessage = new String(Base64.decode(text), "UTF-8");
```

Comment:
Sorry if I didn't make this completely clear on skype, but we shouldn't be calling `String()` on the entire message because there's a good chance that at some point we're going to generate a signature that can't be decoded with UTF-8 and it will raise an unnecessary exception.

I think the easiest way to go about this is to just inspect the raw bytes (iterate over each byte) and look for the third space. Everything after the third space is the binary signature, and everything before it can be cast to a `String` since it's just ascii.